fix(HMSPROV-425): incorporate dejq into the app

lzap opened the pull request:
Since we now have more information and understanding of the domain and environment, the job queue is now Redis (with memory impl for development). This patch moves [dejq](https://github.com/lzap/dejq) code from the external repository into `pkg/worker` and simplifies the API since there is no dbjobqueue anymore. It is as simple as:

```
type Job struct {
	// Random UUID for logging and tracing. It is set by Enqueue function.
	ID uuid.UUID

	// Job type or "queue".
	Type JobType

	// Job arguments.
	Args any
}
```

Jobs are super simple, just args, no output, no return value or error. No dependencies. For this reason, AWS launch job was merged into a single job (with two functions/steps).

The code now uses Go's gob encoding instead of JSON which gives more type safety. Also logging was vastly improvement since there is no need for a logging facade now, all logs now have the following fields:

* account_id
* account_number
* org_id
* job_id
* job_type
* reservation_id

Finally, statistics now also contains "in flight" jobs, meaning jobs which are currently being processed by the worker.

**Edit:** The PR is now ready for full review and testing.

lzap commented on cmd/pbapi/main.go:
```diff
@@ -103,9 +101,9 @@ func main() {
 		log.Fatal().Err(err).Msg("Error initializing dejq queue")
 	}
+	dejq.RegisterJobs(&logger)
 	if config.Worker.Queue == "memory" {
```
We have to register jobs now even in pbapi because registering also performs interface type registration for gob encoder.

lzap commented on deploy/clowdapp.yaml:
```diff
@@ -265,12 +259,4 @@ parameters:
-  - description: Internal queue beat threshold. Now many times the beat can be updated until job is cancelled. Not used in 'redis' mode.
-    name: WORKER_MAX_BEATS
-    value: "10"
+
```
All of these settings were completely removed, I replaced them with:

* Poll interval - how often to poll Redis for new messages (essentially timeout).
* Maximum number of goroutines per worker instance - by default it is CPUs + 1 when unset.

These are now in app configuration, tho, I am not adding them in clowder until we need them. I think defaults are okay.

lzap commented on internal/config/config.go:
```diff
@@ -112,10 +112,9 @@ var config struct {
-		MaxBeats    int           `env:"MAX_BEATS" env-default:"10" env-description:"maximum amount of heartbeats allowed"`
+		Queue        string        `env:"QUEUE" env-default:"memory" env-description:"job worker implementation (memory, redis, sqs, postgres)"`
+		PollInterval time.Duration `env:"POLL_INTERVAL" env-default:"5s" env-description:"polling interval (network timeout)"`
+		MaxPollers   int           `env:"MAX_POLLERS" env-default:"512" env-description:"maximum allowed amount of polling goroutines (CPUs + 1)"`
```
I believe these are sane defaults for all environments.

ezr-ondrej replied:
I'd really love a different name. Poller sounds to me like someone who is performing Pollination, so I immediately think of bees.

What about `MaxThreads`? To me it sounds obvious what the threads would do :shrug:

lzap commented on internal/dao/tests/noop_job_test.go:
```diff
@@ -107,92 +107,6 @@ func TestReservationNoopOneSuccess(t *testing.T) {
-				require.Equal(t, int32(3), updatedRes.Steps)
-				require.Equal(t, "No operation finished", updatedRes.Status)
-			},
-		},
```
The final version does not support steps (dependant jobs), so these tests are irrelevant now.

lzap commented on internal/jobs/common.go:
```diff
@@ -20,10 +20,22 @@ func decodeJob(ctx context.Context, job dejq.Job, argInterface interface{}) erro
-	logger.Info().Interface("args", args).Msgf("Processing job: '%s'", jobName)
 	newContext = ctxval.WithLogger(newContext, &logger)
 	return newContext
 }
```
Every job log now also contain account_number and org_id for better searching in Kibana.

lzap commented on internal/jobs/job_types.go:
```diff
@@ -0,0 +1,9 @@
+	TypeNoop              worker.JobType = "no_operation"
+	TypeLaunchInstanceAws worker.JobType = "launch_instances_aws"
+	TypeLaunchInstanceGcp worker.JobType = "launch_instances_gcp"
+)
```
This was moved from somewhere else.

lzap commented on internal/jobs/launch_instance_aws.go:
```diff
@@ -2,60 +2,162 @@ package jobs
 	// The ARN fetched from Sources which is linked to a specific source
-	ARN *clients.Authentication `json:"arn"`
+	ARN *clients.Authentication
 }
```
Json tags are no longer necessary, the current implementation use gob which is type-safe.

lzap commented on internal/jobs/launch_instance_aws.go:
```diff
@@ -2,60 +2,162 @@ package jobs
 
 	finishJob(ctx, args.ReservationID, jobErr)
-	return jobErr
 }
```
AWS job now have two functions ("steps"):

* Upload public SSH key if not present.
* Launch instance(s).

Error handling is done in this main function.

lzap commented on internal/jobs/launch_instance_aws.go:
```diff
@@ -2,60 +2,162 @@ package jobs
 
 // Job logic, when error is returned the job status is updated accordingly
-func handleLaunchInstanceAWS(ctx context.Context, args *LaunchInstanceAWSTaskArgs) error {
+func DoEnsurePubkeyOnAWS(ctx context.Context, args *LaunchInstanceAWSTaskArgs) error {
```
All functions ("steps") now start with `DoXXX` and are exported so we can write tests in a separate `jobs_test` package (no cycles).

lzap commented on internal/jobs/launch_instance_aws_test.go:
```diff
@@ -18,25 +18,6 @@ import (
-	typed.Region = s.body.Region
-	return nil
-}
-
```
Since job is now a struct and not interface, no need to create stubs.

lzap commented on internal/math/sort.go:
```diff
@@ -0,0 +1,17 @@
+		return b
+	}
+	return a
+}
```
I needed max function, to my surprise, it is not present in the Go stdlib :-D

lzap replied:
Well it would be: `int(math.Min(float64(w.pollMax), float64(runtime.NumCPU()+1)))` which looks awful to be honest, tiny little bit of mathematic code can't hurt any project is what I say :-)

ezr-ondrej replied:
Can't hurt unless there is a bug in it :stuck_out_tongue: every additional line increases the chance of a bug is what I'd say :smile:
But in this case I agree yours looks bit tidier, so I'd be fine with having it, unless @avitova insists :shrug:

lzap commented on internal/queue/dejq/worker.go:
```diff
@@ -0,0 +1,80 @@
+package dejq
```
Oh need to find a better name for this package, there is no `dejq` dependency anymore, will rename in the next push. This is basically the "implementation" package, there is another one called "stub" which is used in unit tests (does nothing).

ezr-ondrej replied:
Why is this separate from the `pkg/worker`? I thought that package could have implementation as well, right?

lzap replied:
Yeah good question, the reason is that I wanted the `pkg/worker` package to be really clean. The `internal/queue` package contains two subpackages (stub and impl) and one or another are imported (in unit tests/in normal code). On top of that, it contains the app-dependent configuration, handler registration and initialization code. This cannot be in the `pkg` package because it depends on internal stuff.

It was your idea to put it into `pkg`, tho, I can move it into `internal` and put everything under one roof: registration of job handlers, configuration, initialization and stub (no operation implementation). Then it could live all in a single package. Honestly, I don't see other project using our very simple job queue and I would probably just copy-paste it if I was starting a new project with the same requirements. So I am all for it, it can be confusing to have two packages: worker and queue.

ezr-ondrej replied:
oki, I'm good with that, it was just an idea, but I do not insist on the `pkg`, you're right that it's not too likely someone would use it :)

My main point here waas: Can we pls rename it from `queue/dejq` to `queue/impl`? :)

lzap replied:
Absolutely, I want the `dejq` name to disappear, it was unpronouncable. What a bad name.

lzap commented on pkg/worker/job.go:
```diff
@@ -0,0 +1,60 @@
+
+	// Number of jobs currently being processed
+	InFlight int64
+}
```
This is the main job queue interface.

lzap commented on pkg/worker/redis.go:
```diff
@@ -0,0 +1,203 @@
+		EnqueuedJobs: uint64(count),
+		InFlight:     w.inFlight,
+	}, nil
+}
```
And this is the "heart" of the stage/prod environments - Redis implementation, uses atomic list operation for the queue, it has a single list for all tasks, each task has a type (string, like a "channel"), UUID (random id for logging, no other purpose) and args which is a generic struct.

It uses gob encoder from the Go stdlib to encode/decode jobs which allows to do type assertion of `any` interface in a safe way (e.g. when a new field is added gob will safely ignore it - leave it blank).

Graceful shutdown is implemented via wait groups, the `Stop` function waits for all workers to finish polling as well as for all goroutines to finish processing of already started jobs.

lzap commented on go.mod:
```diff
@@ -95,21 +94,13 @@ require (
-	github.com/jackc/puddle v1.3.0 // indirect
 	github.com/jackc/puddle/v2 v2.0.0 // indirect
 	github.com/jmespath/go-jmespath v0.4.0 // indirect
 	github.com/joho/godotenv v1.4.0 // indirect
```
Oh forgot to do `go mod tidy` to remove `dejq` from here, will do in my next push.

ezr-ondrej replied:
I still do not see removal of `dejq` :thinking:

lzap replied:
Did not push yet, will do. It will be gone I promise :-)

lzap commented:
Example of trace id not getting propagated into dejq, which this patch will actually solve:

![image](https://user-images.githubusercontent.com/49752/212666630-071a4500-883a-448d-9dbb-af5a86b2a6fa.png)

avitova commented on pkg/worker/redis.go:
```diff
@@ -0,0 +1,203 @@
+func (w *RedisWorker) fetchJob(ctx context.Context) {
+	res, err := w.client.BLPop(ctx, w.pollInterval, w.queueName).Result()
+
+	if errors.Is(err, redis.Nil) {
+		// timeout occurred
+	} else if err != nil {
+		logger := ctxval.Logger(ctx)
+		logger.Error().Err(err).Msg("Error consuming from Redis queue")
+	} else {
+		var job Job
+		dec := gob.NewDecoder(strings.NewReader(res[1]))
+		err = dec.Decode(&job)
+		logger := loggerWithJob(ctx, &job)
+		if err != nil {
+			logger.Error().Err(err).Msg("Unable to unmarshal job payload, skipping")
+		}
+
+		w.handleWG.Add(1)
+		atomic.AddInt64(&w.inFlight, 1)
+		go w.processJob(ctx, &job)
+	}
```
Just a nitpick, but this feels a bit more natural for me.
```suggestion
	if errors.Is(err, redis.Nil) {
		// timeout occurred
		return
	} else if err != nil {
		logger := ctxval.Logger(ctx)
		logger.Error().Err(err).Msg("Error consuming from Redis queue")
		return
	}
	var job Job
	dec := gob.NewDecoder(strings.NewReader(res[1]))
	err = dec.Decode(&job)
	logger := loggerWithJob(ctx, &job)
	if err != nil {
		logger.Error().Err(err).Msg("Unable to unmarshal job payload, skipping")
	}

	w.handleWG.Add(1)
	atomic.AddInt64(&w.inFlight, 1)
	go w.processJob(ctx, &job)
```

lzap replied:
Indeed, because I extracted this from the loop that was the reason it was written this ugly way.

avitova reviewed (commented):
I have noticed one more thing.

ezr-ondrej commented on internal/config/config.go:
```diff
@@ -112,10 +112,9 @@ var config struct {
-		Concurrency int           `env:"CONCURRENCY" env-default:"50" env-description:"number of goroutines handling jobs"`
-		Heartbeat   time.Duration `env:"HEARTBEAT" env-default:"30s" env-description:"heartbeat interval (time interval syntax)"`
-		MaxBeats    int           `env:"MAX_BEATS" env-default:"10" env-description:"maximum amount of heartbeats allowed"`
+		Queue        string        `env:"QUEUE" env-default:"memory" env-description:"job worker implementation (memory, redis, sqs, postgres)"`
```
```suggestion
		Queue        string        `env:"QUEUE" env-default:"memory" env-description:"job worker implementation (memory, redis)"`
```
I guess this is what @adiabramovitch is pointing out ^^

lzap replied:
Yes, I removed intentionally SQS and Postgres, that was the idea.

What is the concern?

ezr-ondrej replied:
The concern is we still have it in the config description as possible value :)

ezr-ondrej commented on internal/jobs/common.go:
```diff
@@ -20,10 +20,22 @@ func decodeJob(ctx context.Context, job dejq.Job, argInterface interface{}) erro
+
 	newContext := ctxval.WithAccountId(ctx, accountId)
-	logger := ctxval.Logger(newContext).With().Int64("reservation_id", reservationId).Logger()
-	logger.Info().Interface("args", args).Msgf("Processing job: '%s'", jobName)
```
I quite liked this line, it was obvious what job has started with what args, why are we removing it? :)

lzap replied:
Because this is actually done here:

```go
func contextLogger(ctx context.Context, accountId, reservationId int64) context.Context {
	logger := ctxval.Logger(ctx).With().
		Int64("reservation_id", reservationId).
		Int64("account_id", accountId).Logger()

	ad := dao.GetAccountDao(ctx)
	account, err := ad.GetById(ctx, accountId)
	if err != nil {
		logger.Warn().Msgf("Unable to fetch account info for: %d", accountId)
	} else {
		logger = logger.With().
			Str("account_number", account.AccountNumber.String).
			Str("org_id", account.OrgID).Logger()
	}

	newContext := ctxval.WithAccountId(ctx, accountId)
	newContext = ctxval.WithLogger(newContext, &logger)
	return newContext
}
```

Every job log should now have all the fields as I describe in the PR description automatically, actually I see few more of these `Int64("reservation_id"` calls which I will remove now.

ezr-ondrej replied:
I'm talking about the first log line at the start of every job, that is now removed, isn't it?

lzap replied:
Perhaps, I am not entirely sure which one you need, my intentions were not to remove anything useful, but I moved some logging messages from service so common functions where they do have more information available, e.g. job_id. Example from api process:

```
10:27AM DBG Started POST request /api/provisioning/v1/reservations/noop binary=pbapi bytes_in=0 hostname=mone.home.lan method=POST remote_ip=127.0.0.1:49619 trace_id=77a9fa786afacb1e602324182fe4e5fa url=/api/provisioning/v1/reservations/noop version=HEAD
10:27AM TRC Account found via org id: 13446659 binary=pbapi bytes_in=0 hostname=mone.home.lan method=POST remote_ip=127.0.0.1:49619 trace_id=77a9fa786afacb1e602324182fe4e5fa url=/api/provisioning/v1/reservations/noop version=HEAD
10:27AM TRC Account cache hit account=3 binary=pbapi bytes_in=0 hostname=mone.home.lan method=POST remote_ip=127.0.0.1:49619 trace_id=77a9fa786afacb1e602324182fe4e5fa url=/api/provisioning/v1/reservations/noop version=HEAD
10:27AM DBG Created a new reservation 30 account_id=3 account_number=6395343 binary=pbapi bytes_in=0 hostname=mone.home.lan method=POST org_id=13446659 remote_ip=127.0.0.1:49619 trace_id=77a9fa786afacb1e602324182fe4e5fa url=/api/provisioning/v1/reservations/noop version=HEAD
10:27AM INF Enqueuing job type: no_operation account_id=3 account_number=6395343 binary=pbapi bytes_in=0 hostname=mone.home.lan job_args={"AccountID":3,"Fail":false,"ReservationID":30} job_id=2003d977-6cf4-406e-a305-54ba3bb2ef0b job_type=no_operation method=POST org_id=13446659 remote_ip=127.0.0.1:49619 trace_id=77a9fa786afacb1e602324182fe4e5fa url=/api/provisioning/v1/reservations/noop version=HEAD
10:27AM INF Pushed job successfully account_id=3 account_number=6395343 binary=pbapi bytes_in=0 hostname=mone.home.lan job_args={"AccountID":3,"Fail":false,"ReservationID":30} job_id=2003d977-6cf4-406e-a305-54ba3bb2ef0b job_result=1 job_type=no_operation method=POST org_id=13446659 remote_ip=127.0.0.1:49619 trace_id=77a9fa786afacb1e602324182fe4e5fa url=/api/provisioning/v1/reservations/noop version=HEAD
10:27AM INF Completed POST request /api/provisioning/v1/reservations/noop in 19ms with 200 binary=pbapi bytes_in=0 bytes_out=22 hostname=mone.home.lan latency_ms=18.738125 method=POST remote_ip=127.0.0.1:49619 status=200 trace_id=77a9fa786afacb1e602324182fe4e5fa url=/api/provisioning/v1/reservations/noop version=HEAD
```

And worker:

```
12:00PM INF Dequeued job from Redis binary=pbworker hostname=mone.home.lan job_args={"AccountID":3,"Fail":false,"ReservationID":31} job_id=7f460d23-9ff0-4471-9ac2-8a576bf21aba job_type=no_operation version=HEAD
12:00PM DBG Reservation status change: 'No operation started' account_id=3 account_number=6395343 binary=pbworker hostname=mone.home.lan job_id=7f460d23-9ff0-4471-9ac2-8a576bf21aba org_id=13446659 reservation_id=31 step=true version=HEAD
12:00PM INF No operation finished account_id=3 account_number=6395343 binary=pbworker hostname=mone.home.lan job_id=7f460d23-9ff0-4471-9ac2-8a576bf21aba org_id=13446659 reservation_id=31 version=HEAD
12:00PM DBG Reservation status change: 'No operation finished' account_id=3 account_number=6395343 binary=pbworker hostname=mone.home.lan job_id=7f460d23-9ff0-4471-9ac2-8a576bf21aba org_id=13446659 reservation_id=31 step=true version=HEAD
12:00PM TRC Increased step number by: 1 account_id=3 account_number=6395343 binary=pbworker hostname=mone.home.lan job_id=7f460d23-9ff0-4471-9ac2-8a576bf21aba org_id=13446659 reservation_id=31 step=true version=HEAD
12:00PM DBG Job step: 1/1 account_id=3 account_number=6395343 binary=pbworker hostname=mone.home.lan job_id=7f460d23-9ff0-4471-9ac2-8a576bf21aba org_id=13446659 reservation_id=31 version=HEAD
12:00PM INF All jobs executed, marking job as success account_id=3 account_number=6395343 binary=pbworker hostname=mone.home.lan job_id=7f460d23-9ff0-4471-9ac2-8a576bf21aba org_id=13446659 reservation_id=31 version=HEAD
```

I feel like logging is now pretty nice, if you miss anything, let me know.

ezr-ondrej reviewed (commented):
Just few nit picks, nothing major, tho I'd put +1 to both @avitova's comments :)

lzap commented on cmd/pbworker/main.go:
```diff
@@ -53,7 +51,6 @@ func main() {
 	logger = logger.With().
 		Timestamp().
 		Str("hostname", hostname).
-		Str("worker_id", xid.New().String()).
```
Oh it was because it is practically useless - in Kibana there are several fields which distinguish pods:

* pod
* hostname
* instance
* others I think too

lzap commented:
Rebased:

* dejq string is gone for the good
* MaxThreads setting
* improved the line of sight
* removed few more "reservation_id" fields because this is now added automatically at the beginning of the job
* I am keeping `pkg/worker` for now, we can move it into `internal` anytime we want

avitova commented on internal/jobs/launch_instance_aws.go:
```diff
@@ -2,60 +2,162 @@ package jobs
+	resDao := dao.GetReservationDao(ctx)
+	awsReservation, err := resDao.GetAWSById(ctx, args.ReservationID)
+	if err != nil {
+		return fmt.Errorf("cannot upload aws pubkey: %w", err)
```
```suggestion
		return fmt.Errorf("cannot get aws reservation by id: %w", err)
```

ezr-ondrej reviewed (approved):
Few nitpicks in the comment threads, but nothing major, I'm good here :+1:
I'll give time to review nitpicks and for others to review, but it can go in as is as far as I'm concerned :)

lzap commented:
Fixed the comment, also noticed during my log review that two messages did not have "job_id" so fixed that as well, here is the diff:

```diff
Subject: [PATCH] fix(HMSPROV-425): incorporate dejq into the app
---
Index: pkg/worker/redis.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pkg/worker/redis.go b/pkg/worker/redis.go
--- a/pkg/worker/redis.go	(revision 1e1d6bd1bcbed452b5a40386e0ec696ca48266db)
+++ b/pkg/worker/redis.go	(date 1674206749236)
@@ -79,9 +79,7 @@
 }
 
 func (w *RedisWorker) Enqueue(ctx context.Context, job *Job) error {
-	logger := loggerWithJob(ctx, job)
 	var err error
-
 	if job.ID == uuid.Nil {
 		job.ID, err = uuid.NewRandom()
 		if err != nil {
@@ -89,6 +87,9 @@
 		}
 	}
 
+	logger := loggerWithJob(ctx, job)
+	logger.Info().Msgf("Enqueuing job type: %s", job.Type)
+
 	var buffer bytes.Buffer
 	enc := gob.NewEncoder(&buffer)
 	err = enc.Encode(&job)
Index: internal/services/noop_reservation_service.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/services/noop_reservation_service.go b/internal/services/noop_reservation_service.go
--- a/internal/services/noop_reservation_service.go	(revision 1e1d6bd1bcbed452b5a40386e0ec696ca48266db)
+++ b/internal/services/noop_reservation_service.go	(date 1674206638727)
@@ -46,7 +46,6 @@
 			ReservationID: reservation.ID,
 		},
 	}
-	logger.Debug().Interface("arg", pj).Msgf("Enqueuing no operation job")
 	err = queue.GetEnqueuer().Enqueue(r.Context(), &pj)
 	if err != nil {
 		renderError(w, r, payloads.NewEnqueueTaskError(r.Context(), "job enqueue error", err))
Index: internal/jobs/common.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/jobs/common.go b/internal/jobs/common.go
--- a/internal/jobs/common.go	(revision 1e1d6bd1bcbed452b5a40386e0ec696ca48266db)
+++ b/internal/jobs/common.go	(date 1674212390403)
@@ -7,10 +7,12 @@
 
 	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
 	"github.com/RHEnVision/provisioning-backend/internal/dao"
+	"github.com/RHEnVision/provisioning-backend/pkg/worker"
 )
 
-func contextLogger(ctx context.Context, accountId, reservationId int64) context.Context {
+func contextLogger(ctx context.Context, job *worker.Job, accountId, reservationId int64) context.Context {
 	logger := ctxval.Logger(ctx).With().
+		Str("job_id", job.ID.String()).
 		Int64("reservation_id", reservationId).
 		Int64("account_id", accountId).Logger()
 
Index: internal/services/gcp_reservation_service.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/services/gcp_reservation_service.go b/internal/services/gcp_reservation_service.go
--- a/internal/services/gcp_reservation_service.go	(revision 1e1d6bd1bcbed452b5a40386e0ec696ca48266db)
+++ b/internal/services/gcp_reservation_service.go	(date 1674206638733)
@@ -114,7 +114,6 @@
 			ProjectID:     authentication,
 		},
 	}
-	logger.Debug().Interface("job", launchJob).Msgf("Enqueuing launch instance job for source %s", reservation.SourceID)
 
 	err = queue.GetEnqueuer().Enqueue(r.Context(), &launchJob)
 	if err != nil {
Index: internal/jobs/noop_job.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/jobs/noop_job.go b/internal/jobs/noop_job.go
--- a/internal/jobs/noop_job.go	(revision 1e1d6bd1bcbed452b5a40386e0ec696ca48266db)
+++ b/internal/jobs/noop_job.go	(date 1674212390399)
@@ -28,7 +28,7 @@
 		return
 	}
 
-	ctx = contextLogger(ctx, args.AccountID, args.ReservationID)
+	ctx = contextLogger(ctx, job, args.AccountID, args.ReservationID)
 
 	jobErr := DoNoop(ctx, &args)
 
Index: internal/jobs/launch_instance_aws.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/jobs/launch_instance_aws.go b/internal/jobs/launch_instance_aws.go
--- a/internal/jobs/launch_instance_aws.go	(revision 1e1d6bd1bcbed452b5a40386e0ec696ca48266db)
+++ b/internal/jobs/launch_instance_aws.go	(date 1674212506301)
@@ -49,7 +49,7 @@
 		return
 	}
 
-	ctx = contextLogger(ctx, args.AccountID, args.ReservationID)
+	ctx = contextLogger(ctx, nil, args.AccountID, args.ReservationID)
 
 	jobErr := DoEnsurePubkeyOnAWS(ctx, &args)
 	if jobErr != nil {
@@ -85,7 +85,7 @@
 	resDao := dao.GetReservationDao(ctx)
 	awsReservation, err := resDao.GetAWSById(ctx, args.ReservationID)
 	if err != nil {
-		return fmt.Errorf("cannot upload aws pubkey: %w", err)
+		return fmt.Errorf("cannot get aws reservation by id: %w", err)
 	}
 
 	pubkey, err := pkDao.GetById(ctx, args.PubkeyID)
Index: internal/jobs/launch_instance_gcp.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/jobs/launch_instance_gcp.go b/internal/jobs/launch_instance_gcp.go
--- a/internal/jobs/launch_instance_gcp.go	(revision 1e1d6bd1bcbed452b5a40386e0ec696ca48266db)
+++ b/internal/jobs/launch_instance_gcp.go	(date 1674212399746)
@@ -45,7 +45,7 @@
 		return
 	}
 
-	ctx = contextLogger(ctx, args.AccountID, args.ReservationID)
+	ctx = contextLogger(ctx, job, args.AccountID, args.ReservationID)
 
 	jobErr := DoLaunchInstanceGCP(ctx, &args)
 
Index: internal/services/aws_reservation_service.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/internal/services/aws_reservation_service.go b/internal/services/aws_reservation_service.go
--- a/internal/services/aws_reservation_service.go	(revision 1e1d6bd1bcbed452b5a40386e0ec696ca48266db)
+++ b/internal/services/aws_reservation_service.go	(date 1674206638731)
@@ -136,7 +136,6 @@
 			ARN:           authentication,
 		},
 	}
-	logger.Debug().Interface("job", launchJob).Msgf("Enqueuing launch instance job for source %s", reservation.SourceID)
 
 	err = queue.GetEnqueuer().Enqueue(r.Context(), &launchJob)
 	if err != nil {
```

lzap commented:
We are good, now, we can take advantage of the fact that background job type is now in our codebase, I propose to:

* Move the accountId from arguments to the job itself, most of our jobs will be account-aware and for those who would not be we can leave zero (will be unused by the job code).
* Add the RH identity in both typed form `identity.XRHID` and as base64 encoded string. Both can be used in jobs as needed (fields from the identity, whole string to be passed as a HTTP/Kafka header).

avitova reviewed (approved):
LGTM!

lzap commented:
Strange, in ephemeral the e2e aws launch test go enqueued and pushed into Redis according to logs, but no workers picked it up.

```
11:11AM INF Enqueuing job type: launch_instances_aws account_id=1 account_number=0369233 binary=pbapi bytes_in=181 hostname=provisioning-backend-api-ff9bc9bfb-9gtcz job_args={"AMI":"ami-0ba22849118c7101a","ARN":{"payload":"arn:aws:iam::093942615996:role/redhat-provisioning-arn-role","type":2},"AccountID":1,"Detail":{"amount":1,"instance_type":"t1.micro","name":"test_DBhdJL0IK8","poweroff":true,"pubkey_name":"","region":"us-east-1"},"PubkeyID":7,"Region":"us-east-1","ReservationID":1,"SourceID":"1"} job_id=99067c8c-edef-4d9e-bb46-b2299f2221bb job_type=launch_instances_aws method=POST org_id=3340851 remote_ip=10.130.11.197:36784 trace_id=75e7b56822e70c8301d150df923eda53 url=/api/provisioning/v1/reservations/aws version=7ce8c4e
11:11AM INF Pushed job successfully account_id=1 account_number=0369233 binary=pbapi bytes_in=181 hostname=provisioning-backend-api-ff9bc9bfb-9gtcz job_args={"AMI":"ami-0ba22849118c7101a","ARN":{"payload":"arn:aws:iam::093942615996:role/redhat-provisioning-arn-role","type":2},"AccountID":1,"Detail":{"amount":1,"instance_type":"t1.micro","name":"test_DBhdJL0IK8","poweroff":true,"pubkey_name":"","region":"us-east-1"},"PubkeyID":7,"Region":"us-east-1","ReservationID":1,"SourceID":"1"} job_id=99067c8c-edef-4d9e-bb46-b2299f2221bb job_result=1 job_type=launch_instances_aws method=POST org_id=3340851 remote_ip=10.130.11.197:36784 trace_id=75e7b56822e70c8301d150df923eda53 url=/api/provisioning/v1/reservations/aws version=7ce8c4e
11:11AM INF Completed POST request /api/provisioning/v1/reservations/aws in 17ms with 200 binary=pbapi bytes_in=181 bytes_out=186 hostname=provisioning-backend-api-ff9bc9bfb-9gtcz latency_ms=17.157652 method=POST remote_ip=10.130.11.197:36784 status=200 trace_id=75e7b56822e70c8301d150df923eda53 url=/api/provisioning/v1/reservations/aws version=7ce8c4e
```

Workers:

```
11:11AM INF Starting 9 polling goroutines binary=pbworker hostname=provisioning-backend-worker-8599c9fcb5-zjszc version=7ce8c4e
```

Both pods have the same config: `Worker:{Queue:redis PollInterval:5s MaxThreads:64}`

lzap commented:
Found it, an accidental `nil` was passed into the `contextLogger` helper function. The GoLand IDE puts `nil` during refactoring when new argument is introduced to a method and AWS provider had `nil`. Nice find IQE! Great stuff.

lzap commented:
Aaand it works! Thanks all. I will do a small followup I noticed during my debugging.

ezr-ondrej commented:
> * Move the accountId from arguments to the job itself, most of our jobs will be account-aware and for those who would not be we can leave zero (will be unused by the job code).
> * Add the RH identity in both typed form `identity.XRHID` and as base64 encoded string. Both can be used in jobs as needed (fields from the identity, whole string to be passed as a HTTP/Kafka header).

All these informations are actually just one, wouldn't it be enough to pass identity.XRHID and have some helpers to get AccountID and the base64 encoded version out of that info? :thinking:

lzap commented:
> All these informations are actually just one, wouldn't it be enough to pass identity.XRHID and have some helpers to get AccountID and the base64 encoded version out of that info?

Yes that is possible as well, assuming that we keep the XRHID structure up to date (e.g. when a field is added). I am proposing this in the middleware cleanup PR as "raw header" I thought it could be useful to just pass it unchanged.
